### PR TITLE
Swarm refactored

### PR DESCRIPTION
**Pull request by elona-dup**

### CODE REVIEW

File: Naive greedy alg/scheduling/Experiments.py [hunk 1]
```diff
@@ -89,3 +59,72 @@ def _generate_job(self, job_count, server_count, num):
             min_num_servers,
             max_num_servers,
         )
+
+    def _get_makespan(self, mass, disparity):
+        mu = numpy.log(mass / disparity)
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Experiments.py]: Prefer using math over numpy
https://devdocs.io/python~3.7/library/math#math.log
https://devdocs.io/python~3.7/library/math#math.sqrt

File: Naive greedy alg/scheduling/Scheduler.py [hunk 2]
```diff
@@ -31,7 +32,9 @@ def random(cls):
         c.shutdown_weight = uniform(0.01, 1.0)
         c.reconfig_prob = uniform(0.001, 1.0)
         c.reconfig_weight = uniform(0.01, 1.0)
-        c.shutdown_time = uniform(370, 1200)
+        c.shutdown_time_1 = uniform(370, 1200)
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Scheduler.py]: Naming needs to be more explicit

File: Naive greedy alg/scheduling/Scheduler.py [hunk 3]
```diff
@@ -133,16 +136,20 @@ def update_schedule(self, time):
...
-                #     random()
-                #     < ((len(av_servers) / len(self.servers)) ** self.conf.shutdown_weight)
-                #     * self.conf.shutdown_prob
-                # ):
-            power_off = Job.make_power_off(
-                [server], start_time=time, duration=self.conf.shutdown_time
-            )
-            self._start_job(power_off)
-            av_servers.remove(server)
+            if (
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Scheduler.py]: For readability, the condition could be wrapped into a function like shutdown_enabled, shutdown_possible.

File: Naive greedy alg/scheduling/Swarm.py [hunk 4]
```diff
@@ -23,14 +24,19 @@ def __init__(self, seed_num, num_particles, num_srvs, num_exp=10):
         self.best_cost = inf
         self.experiment = Experiments()
 
-    def run_epochs(self, num_epochs, draw_stats=False):
+    def run_epochs(self, num_epochs, draw_stats=False, draw_graph=False):
+        epochs_stats = []
         for i in range(num_epochs):
             print("epoch: ", i)
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Swarm.py]: Remove print

File: Naive greedy alg/scheduling/Swarm.py [hunk 5]
```diff
@@ -66,6 +80,12 @@ def _draw_stats(self, num_epoch, particle_idx, stats, config):
         print("Mean cost: ", df_stat["cost"].mean())
         print("Tuning configs:\n", config)
 
+    def _draw_graph(self, num_epoch, stats):
+        visualizer = Visualizer()
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Swarm.py]: The visualizer should not be part of the Swarm. Return instead your list of stats from run_epochs and visualize the result from the main.

File: Naive greedy alg/scheduling/Scheduler.py [hunk 6]
```diff
@@ -74,14 +63,23 @@ def to_dict(self):
...
+        self,
+        server_count,
+        conf,
+        reconfig_enable=True,
+        power_off_enable=True,
+        param_enable=True,
+    ):
         self.servers = [Server(i) for i in range(server_count)]
         self.conf = conf
+        self.reconfig_enable = reconfig_enable
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Scheduler.py]: It should be _enabled.

File: Naive greedy alg/scheduling/Scheduler.py [hunk 7]
```diff
@@ -123,33 +114,47 @@ def update_schedule(self, time):
...
+                job = jobs_by_mass[0]
+                if self._is_job_reconfigurable(job, av_servers, time):
+                    av_servers = self._reconfigure_job(job, av_servers, time)
+                jobs_by_mass.pop(0)
+
+        # Turn off servers
+        if self.power_off_enable:
+            for server in list(av_servers):
+                if self.param_enable:
+                    if not self._shutdown_server(av_servers):
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Scheduler.py]: This should be moved up above if self.param_enable, it might be the cause of the infinite loop.

File: Naive greedy alg/scheduling/Scheduler.py [hunk 8]
```diff
@@ -123,33 +114,47 @@ def update_schedule(self, time):
...
+                        < (
+                            (len(av_servers) / len(self.servers))
+                            ** self.conf.shutdown_weight
+                        )
+                        * self.conf.shutdown_prob
+                    ):
+                        if random() < self.conf.shutdown_time_prob:
+                            shutdown_duration = self.conf.shutdown_time_1
+                        else:
+                            shutdown_duration = self.conf.shutdown_time_2
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Scheduler.py]: We could move this in a separate function and merge the shutdown task creation code
The function could return a boolean (indicating if the server can be shut down) and the appropriate shutdown_duration
This will allow having a clearer workflow.

Smth along these lines:

        # Turn off servers
        if self.power_off_enable:
            for server in list(av_servers):
                if not self._shutdown_server(av_servers):
                    break

                shutdown, duration = self._allow_shutdown(av_servers)
                if shutdown:
                    power_off = Job.make_power_off(
                        [server], start_time=time, duration=duration
                    )
                    self._start_job(power_off)
                    av_servers.remove(server)

File: Naive greedy alg/scheduling/Swarm.py [hunk 9]
```diff
@@ -33,6 +33,12 @@ def run_epochs(self, num_epochs, draw_stats=False, draw_graph=False):
             self._draw_graph(
                 num_epochs, pd.DataFrame([stat.to_dict() for stat in epochs_stats])
             )
+        #   save final best config
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Swarm.py]: This logic does not belong to the Swarm.
Applying the previous comment regarding the visualizer, you can move this logic in the main

File: Naive greedy alg/scheduling/__init__.py [hunk 10]
```diff
@@ -1,14 +1,63 @@
 import logging
+import pandas as pd
 
-from .Swarm import Swarm
+# from .Swarm import Swarm
+from .Experiments import Experiments
+from .Scheduler import SchedulerConfig
+
+# python -m scheduling
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/__init__.py]: To remove (same for the swarm)

File: Naive greedy alg/scheduling/__init__.py [hunk 11]
```diff
@@ -1,14 +1,63 @@
...
-    # swarm.configs()
+    # swarm = Swarm(seed_num=1, num_particles=30, num_srvs=10, num_exp=50)
+    # # swarm.configs()
+    # swarm.run_epochs(num_epochs=100, draw_stats=False, draw_graph=True)
+    # # swarm.configs()
+
+    n_servers = 10
+    n_expts = 30
+    seed = 2
+    # run expt with no reconfig and no power off
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/__init__.py]: I suggest we do 1 function that takes the csv name and **kwargs and then unpack the dictionary when calling run_exps, like run_expts(**kwargs)

File: Naive greedy alg/scheduling/Scheduler.py [hunk 12]
```diff
@@ -195,7 +195,7 @@ def _is_job_reconfigurable(self, job, av_servers, time):
             return False
 
         extra_srv_count = min(job.max_server_count - job.server_count, len(av_servers))
-        if self.param_enable:
+        if extra_srv_count > 0 and self.param_enable:
```
Comment (nesrnesr) [inline Naive greedy alg/scheduling/Scheduler.py]: The added condition should not be necessary as the function is called only if there are available servers and as is_reconfigurable already checks that there is room for at least one more server.

**Review by nesrnesr — CHANGES_REQUESTED** [review]

**Comment by nesrnesr** [comment]
I advise you to first resolve the conflicts

File: Naive greedy alg/scheduling/Experiments.py (hunk 1 above)
Comment (elona-dup) [reply Naive greedy alg/scheduling/Experiments.py]: I don't think it makes much difference as we need to import numpy for the seeding of scipy distribution, but ok.

File: Naive greedy alg/scheduling/Swarm.py (hunk 4 above)
Comment (elona-dup) [reply Naive greedy alg/scheduling/Swarm.py]: I have found the prints pretty useful to monitor how quickly the particles converge and to then decide on the number of epochs to use. I will comment them, if you think that they must be removed, then you can remove them.

File: Naive greedy alg/scheduling/__init__.py (hunk 11 above)
Comment (elona-dup) [reply Naive greedy alg/scheduling/__init__.py]: Seems fine to me. You can implement it if you have time

File: Naive greedy alg/scheduling/Experiments.py (hunk 1 above)
Comment (nesrnesr) [reply Naive greedy alg/scheduling/Experiments.py]: It is preferable to use the most appropriate tools for our needs. It would be restraining. If you use NumPy for everything then you won't be easily able to remove it if you decide not to to use anymore. It doesn't answer a specific need.

File: Naive greedy alg/scheduling/Swarm.py (hunk 4 above)
Comment (nesrnesr) [reply Naive greedy alg/scheduling/Swarm.py]: You can use the Logger instead of the prints, and change its level to enable/disable the prints
